pool.go

```diff
@@ -0,0 +1,172 @@
+package clickhouse
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ClickHouse/clickhouse-go/v2/internal/circular"
+)
+
+type idlePool struct {
+	mu    sync.RWMutex
+	conns *circular.Queue[nativeTransport]
+
+	ticker   *time.Ticker
+	finish   chan struct{}
+	finished chan struct{}
+
+	maxConnLifetime time.Duration
+}
+
+func newIdlePool(lifetime time.Duration, capacity int) *idlePool {
+	pool := &idlePool{
+		conns:           circular.New[nativeTransport](capacity),
+		ticker:          time.NewTicker(lifetime),
+		finish:          make(chan struct{}),
+		finished:        make(chan struct{}),
+		maxConnLifetime: lifetime,
+	}
+
+	go pool.runDrainPool()
+
+	return pool
+}
+
+func (i *idlePool) Len() int {
+	i.mu.RLock()
+	defer i.mu.RUnlock()
+	return i.conns.Len()
+}
+
+func (i *idlePool) Cap() int {
+	i.mu.RLock()
+	defer i.mu.RUnlock()
+	return i.conns.Cap()
+}
+
+func (i *idlePool) Get(ctx context.Context) (nativeTransport, error) {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	if i.closed() {
+		return nil, ErrConnectionClosed
+	}
+
+	for {
```

Comment:
nit: can you put a small comment saying why we have infinite `for` loop here and explain the exit conditions?